# Data Preprocessing 

## Purpose

This module preprocesses high-throughput 10x Genomics Chromium Single Cell  RNA-Seq datasets for subsequent clonal mutational profiling using other [scPhylogenomics analysis modules](https://github.com/ewafula/scPhylogenomics/tree/main/analyses). 

## Analysis
This module preprocesses the raw output from [Cell Ranger count](https://www.10xgenomics.com/support/software/cell-ranger/latest/tutorials/cr-tutorial-ct) for each 10x single cell sample, including barcodes, features, and count matrix to remove contaminating ambient RNA using [SoupX](https://academic.oup.com/gigascience/article/9/12/giaa151/6049831) and doublets/multiplex using [DoubletFinder](https://www.sciencedirect.com/science/article/pii/S2405471219300730). Additionally, low-quality cells often characterized by high mitochondrial RNA are identified using [miQC](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009290) and removed.

#### Module directory structure
Illustration of module directory structure using 10x scRNA-Seq dataset from the scPhylogenomics analysis workflow study. 
```      
.
|-- 01-remove-ambient-rna.R
|-- 02-remove-doublets.R
|-- 03-remove-compromised.R
|-- run-data-preprocessing.sh
|-- README.md
|-- plots
|   |-- AML
|   |   |-- LE1-metrics-distribution.pdf
|   |   |-- LE1-metrics-miqc-filtering.pdf
|   |   |-- LE1-metrics-miqc-model.pdf
|   |   |-- LE1-metrics_correlation.pdf
|   |   |-- LE1-pk-sweep-plot.pdf
|   |   |-- LE1-soupx-contaminant.pdf
|   |   `-- LE1-umap-doublets.pdf
|   `-- TNBC
|       |-- TNBC5-metrics-distribution.pdf
|       |-- TNBC5-metrics-miqc-filtering.pdf
|       |-- TNBC5-metrics-miqc-model.pdf
|       |-- TNBC5-metrics_correlation.pdf
|       |-- TNBC5-pk-sweep-plot.pdf
|       |-- TNBC5-soupx-contaminant.pdf
|       `-- TNBC5-umap-doublets.pdf
|-- results
    |-- AML
    |   `-- LE1-pk-values.txt
    `-- TNBC
        `-- TNBC5-pk-values.txt
```


## General usage of scripts

#### `run-data-preprocessing.sh`
This script serves as a module wrapper script to automate project data preprocessing tasks and can be adapted for any project. All specified paths in this script are relative to the module directory. Therefore, it should always be executed relative to this [module directory](https://github.com/ewafula/scPhylogenomics/tree/main/analyses/data-preprocessing).

###### Example usage:
```
bash run-data-preprocessing.sh
```

#### `01-remove-ambient-rna.R`
This script utilizes the [SoupX](https://github.com/constantAmateur/SoupX) R package to estimate and remove cell-free mRNA contamination (ambient RNA) from the raw output generated by Cell Ranger count across all samples in a project. 

A project's Cell Ranger count sample raw outputs are in the `scPhylogenomics/data/projects/<PROJECT_NAME>/<SAMPLE_NAME>/outs/` sub-directory. The script will create `scPhylogenomics/data/projects/<PROJECT_NAME>/<SAMPLE_NAME>/soupx/` sub-directories containing cell-free mRNA filtered datasets, including barcodes (`barcodes.tsv.gz`), features (`features.tsv.gz`), and a counts matrix (`matrix.mtx.gz`) in each sample directory.

###### Example usage:
```
Rscript --vanilla 01-remove-ambient-rna.R \
    --project TNBC
```

###### Argument descriptions:
```
Usage: 01-remove-ambient-rna.R [options]


Options:
	--project=CHARACTER
		Project directory name with output from 10x Cell Ranger count for all samples

	-h, --help
		Show this help message and exit
```

#### `02-remove-doublets.R`
This script utilizes the [DoubletFinder](https://github.com/chris-mcginnis-ucsf/DoubletFinder) R package to identify and remove doublet and multiplet cells from either Cell Ranger count filtered or SoupX filtered data (barcodes, features, and counts matrices) across all samples in a project. Sample `doublet rates` are obtained from the `Cell Concentration` table in the 10x Chromium Single Cell 3' Reagent Kits user guide associated with the chemistry used for sample preparation in the following resources: 
- [Chromium Next GEM Single Cell 3' Reagent Kits v3.1 (Dual Index)](https://cdn.10xgenomics.com/image/upload/v1722285481/support-documents/CG000315_ChromiumNextGEMSingleCell3__GeneExpression_v3.1_DualIndex__RevF.pdf) 
- [Chromium Next GEM Automated Single Cell 3' Reagent Kits v3.1 (single index)](https://cdn.10xgenomics.com/image/upload/v1686293300/support-documents/CG000286_ChromiumSingleCell3__v3.1_Automation_UG_RevF.pdf)
- [Chromium GEM-X Single Cell 3' Reagent Kits v4](https://cdn.10xgenomics.com/image/upload/v1725314293/support-documents/CG000731_ChromiumGEM-X_SingleCell3v4_UserGuide_RevB.pdf)

A project's Cell Ranger count sample filtered outputs are in the `scPhylogenomics/data/projects/<PROJECT_NAME>/<SAMPLE_NAME>/outs/filtered_feature_bc_matrix` sub-directory. If project samples are filtered to remove ambient RNA, SoupX filtered sample data will be located in the `scPhylogenomics/data/projects/<PROJECT_NAME>/<SAMPLE_NAME>/soupx/` sub-directory. The script will create a `scPhylogenomics/data/projects/<PROJECT_NAME>/<SAMPLE_NAME>/doubletfinder/` sub-directories containing doublets/multiplets filtered datasets, including barcodes (`barcodes.tsv.gz`), features (`features.tsv.gz`), and a counts matrix (`matrix.mtx.gz`) in each sample directory.

###### Example usage:
```
Rscript --vanilla  02-remove-doublets.R \
    --project TNBC \
    --soupx TRUE \
    --components 20 \
    --reagent_kit v3.1_Automated 
```

###### Argument descriptions:
```
Usage: 02-remove-doublets.R [options]


Options:
	--project=CHARACTER
		Project directory name with 10x Cell Ranger counts sample runs

	--soupx
		SoupX filtered feature barcode matrices - [default "TRUE"]
                      If set to FALSE, Cell Ranger filtered feature barcode matrices will be used

	--components=NUMBER
		Number of components for PCA [default 20]

	--reagent_kit=CHARACTER
		10x Chromium Single Cell 3' Reagent Kit used for sample preparation - [default "v3.1_Automated"]
                      - For determining sample doublet rate from recovered cells (v3.1_DualIndex, v3.1_Automated or v4)

	-h, --help
		Show this help message and exit
```

#### `03-remove-compromised.R`
This script utilizes [miQC](https://www.bioconductor.org/packages/release/bioc/vignettes/miQC/inst/doc/miQC.html) implementation within the [SeuratWrappers](https://github.com/satijalab/seurat-wrappers) R package to probabilistically model both the proportion of reads mapped to mitochondrial genes and the total number of detected genes using mixture models to identify and removal of compromised cells from either Cell Ranger count filtered or DoubelFinder filtered data (barcodes, features, and counts matrices) across all samples in a project. In samples where the mixture model only detects a single distribution instead of two (i.e., no significant number of compromised cells), mitochondria content percentile filtering of cells is performed (default = 0.95). Additionally, the script produces QC plots to estimate cutoffs for alternative data approaches.

A project's Cell Ranger count sample filtered data is in the `scPhylogenomics/data/projects/<PROJECT_NAME>/<SAMPLE_NAME>/outs/filtered_feature_bc_matrix` sub-directory. If project samples are filtered to remove doublets/multiplets, the DoubletFinder filtered sample data will be located in the`scPhylogenomics/data/projects/<PROJECT_NAME>/<SAMPLE_NAME>/doubletfinder/` sub-directory. The script will create a `scPhylogenomics/data/projects/<PROJECT_NAME>/<SAMPLE_NAME>/miqc/` sub-directories containing mitochondria filtered datasets, including barcodes (`barcodes.tsv.gz`), features (`features.tsv.gz`), and a counts matrix (`matrix.mtx.gz`) in each sample directory.

###### Example usage:
```
Rscript --vanilla 03-remove-compromised.R \
        --project TNBC \
        --doubletfinder TRUE \
        --min_cells 3 \
        --min_features 200 \
        --percentile_filter 0.95 
```

###### Argument descriptions:
```
Usage: 03-remove-compromised.R [options]


Options:
	--project=CHARACTER
		Project directory name with 10x Cell Ranger counts sample runs

	--doubletfinder
		DoubletFinder filtered feature barcode matrices - [default "TRUE"]
                      If set to FALSE, Cell Ranger filtered feature barcode matrices will be used

	--min_cells=NUMBER
		Minimum features detected in at least this many cells [default 3]

	--min_features=NUMBER
		Minimum cells where at least this many features are detected [default 200]

	--percentile_filter=PERCENTILE
		Filtering when there is no significant number of compromised cells 
                      (i.e., there is only one distribution detected by the mixture model) [default 0.95]

	-h, --help
		Show this help message and exit
```